File: 36-valid-sudoku/valid-sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:

        row_set = defaultdict(set)
        col_set = defaultdict(set)
        box_set = defaultdict(set)

        for i in range(9):
            for j in range(9):

                num = board[i][j]

                if num == ".":
                    continue

                if (num in row_set[i] or num in col_set[j] or num in box_set[(i // 3, j // 3)]):
                    return False

                row_set[i].add(num)
                col_set[j].add(num)
                box_set[(i // 3, j // 3)].add(num)

        return True

File: 36-valid-sudoku/valid-sudoku.java
class Solution {
    public boolean isValidSudoku(char[][] board) {

        Map<Integer, Set<Character>> rs = new HashMap<>();
        Map<Integer, Set<Character>> cs = new HashMap<>();
        Map<Integer, Set<Character>> bs = new HashMap<>();

        for (int i = 0; i < 9; i++) {
            rs.put(i, new HashSet<>());
            cs.put(i, new HashSet<>());
            bs.put(i, new HashSet<>());
        }

        for (int i = 0; i < 9; i++){
            for (int j = 0; j < 9; j++){

                char val = board[i][j];
                if (val == '.'){
                    continue;
                }

                int boxIndex = (i / 3) * 3 + (j / 3);
                
                if (rs.get(i).contains(val) || cs.get(j).contains(val) || bs.get(boxIndex).contains(val)){
                    return false;
                }
                
                rs.get(i).add(val);
                cs.get(j).add(val);
                bs.get(boxIndex).add(val);
            }
        }
        return true;
    }
}